kombu/connection.py

```diff
@@ -170,10 +170,22 @@ def __init__(self, hostname='localhost', userid=None,
                 alt.extend(hostname.split(';'))
                 hostname = alt[0]
             if '+' in hostname[:hostname.index('://')]:
-                # e.g. sqla+mysql://root:masterkey@localhost/
-                params['transport'], params['hostname'] = \
-                    hostname.split('+', 1)
-                self.uri_prefix = params['transport']
+                tran, _, host = hostname.partition('+')
+                # Override the parsed transport with any provided
+                if transport and not isinstance(transport, string_t):
+                    params['transport'] = transport
+                    cls = transport
```

Comment:
Uhm, other parts of the code (IE https://github.com/celery/kombu/pull/859/files#diff-8617041076de8665cfa306742f9b9163L235 ) seem to take for granted that `params['transport']` is a string.

Here it gets set to the provided class, which might have side effects to other parts of the code.

As the transport class doesn't seem to know its name in any way and looking it up in `TRANSPORT_ALIASES` would be fragile for custom transports I'm not sure supporting non-string transports is expected. 
The other branches of the if seem to take for grnated that `transport` is a string and call `get_transport_cls` on it to grab the actual class

Reply:
Running the tests proves otherwise. The test suite (especially the mongo tests) use a passed in transport class to mock out the connection object. NOT using it causes failures when the Connection tries to call out to a real mongodb.

In the original patch set, we bypassed this block if it was mongo, so it went down the other branch (lines 189+). We need a way to use a provided Transport as an override whether or not there's a plus sign. There aren't many transports that support the + schema extension system, so I think we're ok.